deviations/models.py

```diff
@@ -1,15 +1,72 @@
 from datetime import datetime, timedelta
-from typing import Optional
+from typing import TYPE_CHECKING, Generic, Iterable, Optional, TypeVar, Union
 
 from django.urls import reverse
 from django.db import models
 from django.utils.translation import gettext_lazy as _
 
 from exercise.exercise_models import BaseExercise
+from exercise.submission_models import Submission
 from userprofile.models import UserProfile
 from lib.models import UrlMixin
 
 
+TModel = TypeVar('TModel', bound='SubmissionRuleDeviation')
+class SubmissionRuleDeviationManager(models.Manager[TModel], Generic[TModel]):
+    max_order_by: str
+
+    def get_max_deviations(
+        self,
+        submitter: UserProfile,
+        exercises: Iterable[Union[BaseExercise, int]],
+    ) -> Iterable[TModel]:
+        """
+        Returns the maximum deviations for the given submitter in the given
+        exercises (one deviation per exercise is returned). The deviation may
+        be granted to the submitter directly, or to some other submitter in
+        their group.
+        """
+        deviations = (
+            self.filter(
+                models.Q(exercise__in=exercises)
+                & (
+                    # Check that the owner of the deviation is the user, or
+                    # some other user who has submitted the deviation's
+                    # exercise with the user.
+                    models.Q(submitter=submitter)
+                    | models.Exists(
+                        # Note the two 'submitters' filters.
+                        Submission.objects.filter(
+                            exercise=models.OuterRef('exercise'),
+                            submitters=models.OuterRef('submitter'),
+                        ).filter(
+                            submitters=submitter,
+                        )
+                    )
+                )
+            )
+            .select_related('exercise')
+            .order_by('exercise', self.max_order_by)
+        )
+
+        previous_exercise_id = None
+        for deviation in deviations:
+            if deviation.exercise.id == previous_exercise_id:
+                continue
+            previous_exercise_id = deviation.exercise.id
+            yield deviation
+
+    def get_max_deviation(self, submitter: UserProfile, exercise: Union[BaseExercise, int]) -> Optional[TModel]:
+        """
+        Returns the maximum deviation for the given submitter in the given
+        exercise. The deviation may be granted to the submitter directly, or to
+        some other submitter in their group.
+        """
+        deviations = self.get_max_deviations(submitter, [exercise])
+        for deviation in deviations:
+            return deviation
```

Comment:
It's intentional. `get_max_deviation` is a convenience method that takes only one exercise as a parameter and therefore returns only one deviation. It does this by calling `get_max_deviations` and returning the first result. Since `get_max_deviations` is a generator, it doesn't support indexing. I could add a comment to explain this, or I could change that to `return list(deviations)[0]`.

Reply:
It is fine as is. I guess the list conversion would just cost more CPU time. No need for that.